osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs

```diff
@@ -74,6 +74,11 @@ public abstract partial class DrawableHitObject : PoolableDrawableWithLifetime<H
 
         public override bool PropagateNonPositionalInputSubTree => HandleUserInput;
 
+        /// <summary>
+        /// Whether this object should be coloured using its combo position
+        /// </summary>
+        public bool EnableComboColour { get; set; } = true;
```

Comment:
I'd like to say this flag shouldn't be required to make this work... What is the reason for adding it?

Reply:
When testing, for some reason the snap coloring would only apply to the first few notes of a song. After that, it would revert to the original combo colors with occasional off-colors. Here are two captures I took without the flag, the first showing the snap color ending, and the second showing off-colors after the snap coloring ends:


https://github.com/ppy/osu/assets/1653385/393255b2-1dcd-487b-9d56-bb26713065f8


https://github.com/ppy/osu/assets/1653385/b190d2cc-d8e2-4167-ade3-11785da64eec

Reply:
I looked into the issue more, and I think I narrowed it down: There's an additional call to UpdateComboColour at the end of OnApply in DrawableHitObject on line 283 with the following comment:
``` 
// Combo colour may have been applied via a bindable flow while no object entry was attached.
// Update here to ensure we're in a good state.
UpdateComboColour();
```

This is only called after the first few objects in the map. All objects, including the first objects in the map, have UpdateComboColour called from the bindings in LoadComplete around line 205:
```
comboIndexBindable.BindValueChanged(_ => UpdateComboColour());
comboIndexWithOffsetsBindable.BindValueChanged(_ => UpdateComboColour(), true);
```

Only these calls are overridden by changing the color in ApplyCustomUpdateState. If I comment out the UpdateComboColour() on line 283 above, the snap coloring works for the whole map without the flag added (but I have yet do do enough testing to see what other side effects this has.)

Reply:
Thanks for the explanation.

I tried to see if this is fixable by overriding the combo colour on `drawableObject.HitObjectApplied` too, but unfortunately it doesn't do the trick (rewinding replays, or changing combo colour normalisation break that). It appears that the combo colour is too deeply coupled into everything else to be changed reliably via existing hooks.

We aim for maximal separation of mods from game logic, so I would like to see this flag gone.

Possible ways out to explore:
- Revert back to overriding the colour in `Update()`
- Add some special interface for this goal that can be used to hook into the `UpdateComboColour()` flow (`IApplicableToComboColours`? `IApplicableToSkin` maybe, even)?

Due to the number of other open pulls and/or lack of time I am not able to put time in to test these approaches out. Maybe we could be fine with the crude `Update()` thing for now. Dunno.

Reply:
I'd be okay with `Update` as a short term solution. Beats having one more PR open indefinitely.

Reply:
I changed it back to using OnUpdate and removed the changes to DrawableHitObject for now.